src/codegen/compile.ml

```diff
@@ -1816,111 +1831,20 @@ module BitTagged = struct
-            let lower_bound = Int32.(neg (shift_left 1l sbits)) in
-            let upper_bound = Int32.shift_left 1l sbits in
-            (* lower_bound <= x < upper_bound *)
-            compile_unboxed_const lower_bound ^^
-            get_x ^^
-            G.i (Compare (Wasm.Values.I32 I32Op.LeS)) ^^
-            get_x ^^ compile_unboxed_const upper_bound ^^
-            G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^
-            G.i (Binary (Wasm.Values.I32 I32Op.And))
-         | Nat64 | Nat32 ->
-            let ubits = ubits_of pty in
-            let upper_bound = Int32.shift_left 1l ubits in
-            (* 0 <= x < upper_bound *)
-            get_x ^^ compile_unboxed_const upper_bound ^^
-            G.i (Compare (Wasm.Values.I32 I32Op.LtU))
-         | _ ->
-            assert false) ^^
-         get_res ^^
-         G.i (Compare (Wasm.Values.I32 I32Op.Eq)) ^^
-         E.else_trap_with env (prim_fun_name pty "check_can_tag_i32") ^^
-         get_res)
-    else
-      G.nop
-
-  let if_can_tag_i32 env pty retty is1 is2 = Type.(match pty with
-    | Nat | Int | Int64 | Int32 ->
-      Func.share_code1 Func.Never env
-        (prim_fun_name pty "if_can_tag_i32") ("x", I32Type) [I32Type] (fun env get_x ->
-          (* checks that all but the low sbits are both either 0 or 1 *)
-          get_x ^^
-          get_x ^^ compile_shrS_const (Int32.of_int (32 - sbits_of pty)) ^^
-          G.i (Binary (Wasm.Values.I32 I32Op.Xor)) ^^
-          compile_shrU_const (Int32.of_int (sbits_of pty)) ^^
-          G.i (Test (Wasm.Values.I32 I32Op.Eqz)) ^^
-          sanity_check_can_tag_i32 env pty get_x)
-      ^^
-      E.if_ env retty is1 is2
-    | Nat64 | Nat32 ->
-       Func.share_code1 Func.Never env
-         (prim_fun_name pty "if_can_tag_i32") ("x", I32Type) [I32Type] (fun env get_x ->
-          (* checks that all but the low ubits are 0 *)
-          get_x ^^ compile_shrU_const (Int32.of_int (ubits_of pty)) ^^
-          G.i (Test (Wasm.Values.I32 I32Op.Eqz)) ^^
-          sanity_check_can_tag_i32 env pty get_x)
-      ^^
-      E.if_ env retty is1 is2
-    | _ -> assert false)
-
-  let if_can_tag_u32 env pty retty is1 is2 = Type.(
-    match pty with
-    | Nat | Int | Int64 | Int32 ->
-      let sbits = sbits_of pty in
-      compile_shrU_const (Int32.of_int sbits) ^^
-      E.if_ env retty is2 is1 (* NB: swapped branches *)
-    | Nat64 | Nat32 ->
-      let ubits = ubits_of pty in
-      compile_shrU_const (Int32.of_int ubits) ^^
-      E.if_ env retty is2 is1 (* NB: swapped branches *)
-    | _ -> assert false)
-
-  let tag_i32 env pty =
-    let ubits = ubits_of pty in
-    compile_shl_const (Int32.sub 32l (Int32.of_int ubits)) ^^
-    (* tag *)
-    compile_bitor_const (TaggingScheme.tag_of_typ pty)
-
-  let untag_i32 line env pty = Type.(match pty with
-    | Nat | Int | Int64 | Int32 ->
-      let ubits = ubits_of pty in
-      (* check tag *)
+    | Nat | Int | Int64 | Int32 | Int16 | Int8 ->
```

Comment:
Ah, ok, the extra case for Int16 and Int8 are because we no longer use `UnboxedWord32` for these

Reply:
Yes, actually, as you already assumed, I removed `UnboxedWord32` even before the precise tagging PR. So, already, some history in this PR...